mlir/lib/Catalyst/Transforms/SplitMultipleTapes.cpp

```diff
@@ -0,0 +1,307 @@
+        for (Value v : RetValues) {
+            RetTypes.push_back(v.getType());
+        }
+
+        // 2. Create a scf.execute_region and wrap it around the tape ops
+        // The scf.execute_region needs to yield (aka return) the values identified
+        std::pair<scf::ExecuteRegionOp, scf::YieldOp> WrappingRegionAndYield =
+            wrapTapeOpsInSCFRegion(TapeOps, RetValues, RetTypes, builder,
+                                   OriginalMultitapeFunc->getLoc());
+        scf::ExecuteRegionOp executeRegionOp = WrappingRegionAndYield.first;
+        scf::YieldOp SCFRegionYieldOp = WrappingRegionAndYield.second;
+
+        // 3. The scf outliner automatically captures values not defined in the
+        // to-be-outlined function body and transforms them to the outlined function's
+        // arguments, but it does not automatically propagate the returned value
+        // downstream to replace the previously used values. This we need to do manually.
+
+        // At this stage, the scf.yield values ARE the ones to be replaced by
+        // the scf region's result values, since we never replaced anything yet
+        // Of course, don't replace the use in the region itself!
+        propagateSCFRetValsDownstream(executeRegionOp, SCFRegionYieldOp, TapeOps, RetValues);
+
+        // 4. Outline the region
+        func::CallOp call;
+        FailureOr<func::FuncOp> outlined = outlineSingleBlockRegion(
+            builder, OriginalMultitapeFunc->getLoc(), executeRegionOp.getRegion(),
+            OriginalMultitapeFunc.getSymName().str() + "_tape_" + std::to_string(tape_number),
+            &call);
+        OutlinedFuncs.push_back(outlined);
+
+        if (failed(outlined)) {
+            return failure();
+        }
+        return success();
+    } // createTapeFunction()
+
+    void runOnOperation() override
+    {
+        Operation *module = getOperation();
+        mlir::IRRewriter builder(module->getContext());
+
+        // 1. Identify the functions with multiple tapes
+        // Walk through each function and count the number of devices.
+        // In frontend when tracing (jax_tracers.py/trace_quantum_function),
+        // each tape has its own quantum.device operation attached to it
+        SmallVector<func::FuncOp> MultitapePrograms;
+        module->walk([&](func::FuncOp func) {
+            if (countTapes(func) >= 2) {
+                MultitapePrograms.push_back(func);
+            }
+        });
+
+        // Do nothing and exit for classical and single-tape programs
+        if (MultitapePrograms.empty()) {
+            return;
+        }
+
+        // 2. Parse the function into tapes
+        // total number of operation lists is number of tapes +2
+        // for classical pre and post processing
+        for (func::FuncOp func : MultitapePrograms) {
+            SmallVector<std::shared_ptr<SmallVector<Operation *>>> OpsEachTape(countTapes(func) + 2,
```

Comment:
You were asking about naming conventions earlier:
- variables and functions: lowerCaseCamel
- classes: UpperCaseCamel

I think there is a bit of a mixture in this code, and I also saw snake_case further up which we should avoid.